### ISSUE TITLE
[WIP] On-the-fly log-abs-det-Jacobian and caching intermediate values

### ISSUE THREAD

**Issue by vmoens** [facebookincubator/flowtorch#88]
Currently, computing the log absolute determinant of the Jacobian (LADJ) is achieved via an explicit inversion of the flow result (say y) onto the input domain (say x). The gradient of the LADJ is used for the loss, hence every operation (forward and backward through the flow) must be differentiable.
WLOG If the LADJ is represented by the function `h` and the flow by the function `f` with inverse `g`, this amends to computing `h(g(f(x)))` whose gradient can trivially be shown to be equal to the gradient of `h(x)`.
The advantage of this method is that we can avoid storing intermediate values. It also paves the way for backpropagation [without storing intermediate values](https://proceedings.neurips.cc/paper/2017/file/f9be311e65d81a9ad8150a60844bb94c-Paper.pdf).

However, this approach relies on the following implicit assumptions:
1. computing the inverse can be achieved with an arbitrary precision tolerance
2. computing the inverse is cheap
3. the inverse operation `g` is differentiable

Several architectures violate these assumptions, at least from first principle: 
- Invertible residual networks have no analytical inverse and require a solver to compute `g`. Gradients can be passed through this inversion but not at no cost. During training, there may be points in time where the inverse is harder to gather, for instance because the spectral normalization has not been completed (too few steps). 
- In general, "rich" and meaningful transforms (i.e. with a high Lipschitz constant) will have poor inverses. See [this](https://arxiv.org/pdf/2006.09347.pdf) for a theoretical approach. 
- Some types of flows are not "truly" invertible: see [CIF](https://arxiv.org/pdf/2007.05426.pdf) and [SurVAE](https://www.google.com/search?q=SurVAE+Flows).
- In most cases, the inversion will be computationally equally or more involved than the forward transform. 

To add to these limitations, computing the inverse transform on top of the forward transform occupies twice more space in the memory than just caching intermediate values unless backpropagation without storing intermediate values is used. For deep models (e.g. [Glow](https://arxiv.org/pdf/1807.03039v2.pdf)) this will limit FlowTorch capacity to model complex distributions.

## Proposals
The obvious solution is to compute the LADJ on-the-fly, during the forward transform. 
However, the LADJ computation does not usually come for free unless in specific cases (e.g. coupling, householder and other flows).
The following solutions permit to compute the LADJ at a reduced cost and with greater precision.

1. **Context manager**:
This is not an independent solution _per se_ but I write it here as a reminder that we could use this logic to turn on/off downstream features with this kind of tool.
In PyTorch, context managers can be used to activate/deactivate graph computation using `torch.enable_grad()` or `torch.no_grad()`. A similar context manager could be used to compute the LADJ, or to indicate that some intermediate operations should be executed at run time to allow for an efficient computation of the LADJ.

2. **Caching intermediate results**:
We could cache the input and output in a cache for future use:
```python
class Bijector(metaclass=flowtorch.LazyMeta):
    def __init__(self, ..., cache_size=1):
        ...
        self._cache = self._make_cache(cache_size)

    def _make_cache(self, cache_size: int) -> List[dict]:
        caches = []
        for i in range(cache_size):
            caches.append({
                "x": None,
                "y": None,
            })
        return caches
```
During execution, we could store the input and output at each step of the flow with some kind of decorator:
```python
    def forward(
        self,
        x: torch.Tensor,
        context: Optional[torch.Tensor] = None,
    ) -> torch.Tensor:
        # TODO: Allow that context can have a batch shape
        assert context is None  # or context.shape == (self._context_size,)
        for _cache in self._cache:
            if x is _cache['x']:
                return _cache['y']
        y = self._forward(x, context)
        if len(self._cache):
            _cache = self._erase_cache(self._cache[-1])
            _cache["x"] = x
            _cache["y"] = y
            _cache["mode"] = "forward"
            self._cache = [_cache, *self._cache[1:]]
        return y
```
When calling `inverse`, the output `y` is checked against the value stored in the cache. If the output id matches the previous output id, then the respective input is returned. This avoids a potentially burdensome operation. If it doesn't match, the cache is erased (or the first cache of the list is erased, if the cache can handle multiple calls) and filled with the new value. We must store the "mode", being either `"forward"` or `"backward"`, to make sure that we don't compute the LADJ on the value resulting from a backward call on the flow.

```python
    def inverse(
        self,
        y: torch.Tensor,
        x: Optional[torch.Tensor] = None,
        context: Optional[torch.Tensor] = None,
    ) -> torch.Tensor:
        # TODO: Allow that context can have a batch shape
        assert context is None  # or context.shape == (self._context_size,)
        for _cache in self._cache:
            if y is _cache['y']:
                return _cache['x']
        return self._inverse(y, x, context)
```

There are drawbacks to this method: for instance, we would need to choose what to do when the flow is copied, serialized, when the dtype or the device is changed (erase the cache?).

3. **Tensor method**
Another option could be for the class `Bijector` to return a `torch.Tensor`-like object that tracks the operations executed in the flow:
```python
from __future__ import annotations
import torch
from typing import Union

class BijectiveTensor():
    def __init__(self, tensor: torch.Tensor, originating_tensor: Union[torch.Tensor, BijectiveTensor] , transform: Bijector):
        self._register_originating_tensor(originating_tensor)
        self.transform = transform
        self._tensor = tensor

    def _register_originating_tensor(self, other: Union[torch.Tensor, BijectiveTensor]):
        self.originating_tensor = other

    def _check_is_originating(self, other):
        if self.originating_tensor is other: 
            return True
        if not isinstance(self.originating_tensor, BijectiveTensor):
            return self.originating_tensor is other
        else:
            return self.originating_tensor._check_is_originating(other)

    def log_abs_det_jacobian(self, other: torch.Tensor):
        if not self._check_is_originating(other):
            raise RuntimeError("Originating tensor has not been found to be part of the graph of the BijectiveTensor")
        ldj = self.transform.log_abs_det_jacobian(self.originating_tensor, self)
        if other is not self.originating_tensor:
            return self.originating_tensor.log_abs_det_jacobian(other) + ldj
        return ldj
    
    @classmethod
    def __torch_function__(cls, func, types, args=(), kwargs=None):
        if kwargs is None:
            kwargs = {}
        args = [a._tensor if hasattr(a, '_tensor') else a for a in args]
        ret = func(*args, **kwargs)
        return ret
    
    def __add__(self, other):
        return self._tensor+other
    
    def __mul__(self, other):
        return self._tensor*other
    
    def __truediv__(self, other):
        return self._tensor/other

    def __sub__(self, other):
        return self._tensor-other
```

A concrete toy example could be
```python
class BijectorSum:
    def forward(self, x):
        y = x+2
        return BijectiveTensor(y, x, self)
    
    def log_abs_det_jacobian(self, x, y):
        return torch.zeros_like(x)
class BijectorMul:
    def forward(self, x):
        y = x*2
        return BijectiveTensor(y, x, self)
    
    def log_abs_det_jacobian(self, x, y):
        return (torch.ones_like(x)*2).log()
    
    
x = torch.randn(3)
b1 = BijectorSum()
b2 = BijectorMul()
x1 = b1.forward(x)
y = b2.forward(x1)

y.log_abs_det_jacobian(x) # LADJ of the whole chain
y.log_abs_det_jacobian(x1) # LADJ of the last operation
```

Coming back to the first point of this proposal, we could also have a context manager class that controls whether the flow output should be a `BijectiveTensor` or a regular `torch.Tensor` (to spare some memory for instance).

**Comment by stefanwebb** [comment]
@vmoens thanks for the detailed proposal - this is fab!

Approach 2 is what we did in Pyro and what I was thinking we'd do for FlowTorch, although Approach 3 has piqued my curiosity.

Let's chat next week offline!

**Comment by vmoens** [comment]
Closes as from #89
